File: experiments/006-SmalltalkVsJava/cells/java-gradle_1-Empty_free/2024-1c-parcial1/20260915-112708-41868/analysis/own-tests-project/src/main/java/simshotels/Hotel.java
package simshotels;

import java.util.List;
import java.util.function.ToIntFunction;

public class Hotel {

    private final List<Floor> floors;

    // instance creation (class side)

    public static Hotel with(List<Floor> aFloorsCollection) {
        Hotel.assertHaveFloors(aFloorsCollection);

        return new Hotel(aFloorsCollection);
    }

    // initialization

    private Hotel(List<Floor> aFloorsCollection) {
        floors = aFloorsCollection;
    }

    // guests

    public void receiveAtFloorAtRoom(String aGuestType, int aFloorNumber, int aRoomNumber) {
        floorAt(aFloorNumber).receiveAtRoom(aGuestType, aRoomNumber);
    }

    public void receiveWithReservationAtFloorAtRoom(String aGuestType, int aFloorNumber, int aRoomNumber) {
        floorAt(aFloorNumber).receiveWithReservationAtRoom(aGuestType, aRoomNumber);
    }

    public void reserveRoomAtFloor(int aRoomNumber, int aFloorNumber) {
        floorAt(aFloorNumber).reserveRoom(aRoomNumber);
    }

    // accounting

    public int totalLosses() {
        return totalOf(Floor::totalLosses);
    }

    public int totalProfits() {
        return totalOf(Floor::totalProfits);
    }

    // totals

    public int totalRooms() {
        return totalOf(Floor::totalRooms);
    }

    public int totalRoomsAvailable() {
        return totalOf(Floor::totalRoomsAvailable);
    }

    public int totalRoomsOccupied() {
        return totalOf(Floor::totalRoomsOccupied);
    }

    public int totalRoomsReserved() {
        return totalOf(Floor::totalRoomsReserved);
    }

    // testing

    public boolean isEmpty() {
        return floors.stream().allMatch(Floor::isAvailable);
    }

    // accessing

    private Floor floorAt(int aFloorNumber) {
        if (aFloorNumber < 1 || aFloorNumber > floors.size()) {
            Hotel.signalFloorNumberDoesNotExist();
        }

        return floors.get(aFloorNumber - 1);
    }

    private int totalOf(ToIntFunction<Floor> aFloorTotal) {
        return floors.stream().mapToInt(aFloorTotal).sum();
    }

    // assertions (class side)

    public static void assertHaveFloors(List<Floor> aFloorsCollection) {
        if (aFloorsCollection.isEmpty()) {
            Hotel.signalNoFloors();
        }
    }

    // error description (class side)

    public static String floorNumberDoesNotExistErrorDescription() {
        return "Floor number does not exist";
    }

    public static String noFloorsErrorDescription() {
        return "Cannot have a Hotel without floors";
    }

    public static String roomNumberDoesNotExistErrorDescription() {
        return Floor.roomNumberDoesNotExistErrorDescription();
    }

    // exceptions (class side)

    public static void signalFloorNumberDoesNotExist() {
        throw new RuntimeException(Hotel.floorNumberDoesNotExistErrorDescription());
    }

    public static void signalNoFloors() {
        throw new RuntimeException(Hotel.noFloorsErrorDescription());
    }
}

File: experiments/020-PythonVsJava/cells/python-pytest_1-Empty_free/2024-1c-parcial1/20260921-211055-3645/analysis/own-tests-project/src/simshotels/hotel.py
from simshotels.floor import Floor


class Hotel:

    # instance creation (class side)

    @classmethod
    def with_floors(cls, a_floors_collection):
        Hotel.assert_have_floors(a_floors_collection)

        return cls(a_floors_collection)

    # initialization

    def __init__(self, a_floors_collection):
        self._floors = a_floors_collection

    # testing

    def is_empty(self):
        return all(floor.is_available() for floor in self._floors)

    # accounting

    def total_losses(self):
        return self.total_of(lambda floor: floor.total_losses())

    def total_profits(self):
        return self.total_of(lambda floor: floor.total_profits())

    # totals

    def total_rooms(self):
        return self.total_of(lambda floor: floor.total_rooms())

    def total_rooms_available(self):
        return self.total_of(lambda floor: floor.total_rooms_available())

    def total_rooms_occupied(self):
        return self.total_of(lambda floor: floor.total_rooms_occupied())

    def total_rooms_reserved(self):
        return self.total_of(lambda floor: floor.total_rooms_reserved())

    def total_of(self, a_floor_total):
        return sum(a_floor_total(floor) for floor in self._floors)

    # guests

    def receive_at_floor_at_room(self, a_guest_type, a_floor_number, a_room_number):
        self.floor_at(a_floor_number).receive_at_room(a_guest_type, a_room_number)

    def receive_with_reservation_at_floor_at_room(self, a_guest_type, a_floor_number, a_room_number):
        self.floor_at(a_floor_number).receive_with_reservation_at_room(a_guest_type, a_room_number)

    def reserve_room_at_floor(self, a_room_number, a_floor_number):
        self.floor_at(a_floor_number).reserve_room(a_room_number)

    # accessing floors

    def floor_at(self, a_floor_number):
        if a_floor_number < 1 or a_floor_number > len(self._floors):
            Hotel.signal_floor_number_does_not_exist()

        return self._floors[a_floor_number - 1]

    # assertions (class side)

    @staticmethod
    def assert_have_floors(a_floors_collection):
        if len(a_floors_collection) == 0:
            Hotel.signal_no_floors()

    # error description (class side)

    @staticmethod
    def floor_number_does_not_exist_error_description():
        return "Floor number does not exist"

    @staticmethod
    def no_floors_error_description():
        return "Cannot have a Hotel without floors"

    # the rooms of a floor are the floor's responsibility, so is the error
    @staticmethod
    def room_number_does_not_exist_error_description():
        return Floor.room_number_does_not_exist_error_description()

    # exceptions (class side)

    @staticmethod
    def signal_floor_number_does_not_exist():
        raise RuntimeError(Hotel.floor_number_does_not_exist_error_description())

    @staticmethod
    def signal_no_floors():
        raise RuntimeError(Hotel.no_floors_error_description())
